sdks/go/pkg/beam/core/runtime/metricsx/metricsx.go

```diff
@@ -78,17 +78,54 @@ func groupByType(minfos []*pipepb.MonitoringInfo) (
 				continue
 			}
 			gauges[key] = value
-		case
-			UrnToString(UrnStartBundle),
-			UrnToString(UrnProcessBundle),
-			UrnToString(UrnFinishBundle),
-			UrnToString(UrnTransformTotalTime):
+		case UrnToString(UrnStartBundle):
 			value, err := extractMsecValue(r)
 			if err != nil {
 				log.Println(err)
 				continue
 			}
-			msecs[key] = value
+			if v, ok := msecs[key]; ok {
+				v.Start = value
+				msecs[key] = v
+			} else {
+				msecs[key] = metrics.MsecValue{Start: value}
+			}
+		case UrnToString(UrnProcessBundle):
+			value, err := extractMsecValue(r)
+			if err != nil {
+				log.Println(err)
+				continue
+			}
+			if v, ok := msecs[key]; ok {
+				v.Process = value
+				msecs[key] = v
+			} else {
+				msecs[key] = metrics.MsecValue{Process: value}
+			}
+		case UrnToString(UrnFinishBundle):
+			value, err := extractMsecValue(r)
+			if err != nil {
+				log.Println(err)
+				continue
+			}
+			if v, ok := msecs[key]; ok {
+				v.Finish = value
+				msecs[key] = v
+			} else {
+				msecs[key] = metrics.MsecValue{Finish: value}
+			}
+		case UrnToString(UrnTransformTotalTime):
+			value, err := extractMsecValue(r)
+			if err != nil {
+				log.Println(err)
+				continue
+			}
+			if v, ok := msecs[key]; ok {
+				v.Total = value
+				msecs[key] = v
+			} else {
+				msecs[key] = metrics.MsecValue{Total: value}
+			}
```

Comment:
Instead of duplicating the code out, consider instead adding a single inner switch statement to get to the right field to set or re-set. There's no need for all this boilerplate repetition.
eg.
1. get or create value. (note, that map values, if they aren't in a map, are already a 0 value.
2. Switch statement to update the field appropriately.
3. set the entry in the map to the updated value.
//